app/api/chat/route.ts

```diff
@@ -1,44 +1,160 @@
 import { NextResponse } from "next/server";
-import { getChatById, saveChat, saveMessage } from "../../../lib/db/queries";
-import { auth } from "../../auth";
-import { createDataStreamResponse, Message, streamText } from "ai";
-import { groq } from "@ai-sdk/groq";
-import { Chat } from "../../../lib/db/schema";
+import {
+  getChatById,
+  getPredictionByMarketAddress,
+  saveChat,
+  saveMessage,
+  savePrediction,
+} from "@/lib/db/queries";
+import {
+  createDataStreamResponse,
+  generateText,
+  Message,
+  streamText,
+} from "ai";
+import { Chat } from "@/lib/db/schema";
+import { generateSystemPrompt, jsonPrompt } from "../prompt";
+import { isAddress } from "viem";
+import { getIronSession } from "iron-session";
+import { SessionData, sessionOptions } from "@/lib/session";
+import { cookies } from "next/headers";
+import { groqModel } from "@/lib/ai/groq";
 
 export async function POST(request: Request) {
-  const { message }: { message: string } = await request.json();
-  const session = await auth();
+  const { message, marketId }: { message: string; marketId: string } =
+    await request.json();
+  const session = await getIronSession<SessionData>(
+    await cookies(),
+    sessionOptions
+  );
 
-  if (!session || !session.user || !session.user.id) {
+  if (!session || !session.userId) {
     return new Response("Unauthorized", { status: 401 });
   }
 
   const title =
     message.length > 40 ? message.slice(0, 40).concat("...") : message;
 
-  const saveChatResult = await saveChat({
-    userId: session.user.id,
+  const [chat] = await saveChat({
+    userId: session.userId,
     title,
+    marketAddress: marketId,
   });
-  const chat = saveChatResult[0];
 
   await saveMessage({
     chatId: chat.id,
     message: { response: message },
     role: "user",
   });
 
+  try {
+    if (marketId) savePredictionAnswer(message, chat, marketId);
+    else verifyQuestionAndSaveAnswer(message, chat);
+  } catch (error) {
+    let message;
+    if (error instanceof Error) message = error.message;
+    else message = String(error);
+
+    return new Response(message, { status: 400 });
+  }
+
   return NextResponse.json({ chatId: chat.id });
 }
 
+async function savePredictionAnswer(
+  message: string,
+  chat: Chat,
+  marketAddress: string
+) {
+  const prediction = await getPredictionByMarketAddress({ marketAddress });
+
+  if (!prediction) {
+    const predicitonText = await generatePrediction(
+      message,
+      chat,
+      marketAddress
+    );
+    await savePrediction({
+      content: predicitonText,
+      marketAddress: marketAddress,
+    });
+    return;
+  }
+
+  await saveMessage({
+    chatId: chat.id,
+    message: {
+      response: prediction.content as string,
+    },
+    role: "assistant",
```

Comment:
haven't we saved chat on `generatePrediction`?

Reply:
no, it is just saving the prediction message on the chat, it was a wrong naming